ops/log.py

```diff
@@ -53,6 +55,21 @@ def setup_root_logging(
         debug: if True, write logs to stderr as well as to juju-log.
         exc_stderr: if True, write uncaught exceptions to stderr as well as to juju-log.
     """
+    logging.captureWarnings(True)
```

Comment:
Something I haven't thought through (maybe you already have?) is what happens when testing. Without this change, would the warnings end up in the pytest output? With the change, do they vanish because juju debug-log output isn't displayed? Are Harness and Scenario different here? I think probably they would still show in Scenario.

Actually, this is related to what you asked me about the test failing when you run the whole suite. I guess we first want to figure out what the desired behaviour is. Maybe a good team discussion?

Reply:
Without `logging.captureWarnings(True)`, the warning shows in pytest's output, and the test case fails, because warnings aren't redirectedted, so it's still in the output, not in logging, so `JujuLogHandler` can't catch it.

And yes, with the change, the warnings won't show up in the output because this line essentially redirects warnings to the logging system.

Reply:
[In the original issue](https://github.com/canonical/operator/issues/1077#issuecomment-2136815485) I mentioned two methods.

One is in the draft PR, which if se use logging.captureWarnings(), warnings won't show up in the output any more, because warnings are redirected to the logging system.

The other way I mentioned is showwarning(), defaults to stderr, if we want to keep warnings in the output and in the logging at the same time, we can do it this way, log it, then print it.

To me, I think keeping the warnings in one place is better, no duplication. But maybe in the stderr is better than in juju debug log, because users expect warnings to be in stderr, and users may have centralized logging systems and it'd be hard for them to figure out why warnings don't show up there.

Which behaviour do we want? @benhoyt Let's have a quick discussion in our next daily.